### ISSUE TITLE
bind/shortcut for search?

### ISSUE THREAD

**Issue by gliptak** [lusakasa/saka-key#25]
Could a bind for search be added? In maybe bound to ```/```? Thanks

**Comment by eejdoowad** [comment]
This is a feature I think would be extremely useful. I personally want a find mode similar to what VSCode provides:

![image](https://user-images.githubusercontent.com/2775718/29760667-bcb5dd9e-8b79-11e7-8978-846718f58386.png)

I'm not sure when I'll get a chance to work on it. The Todos list grows, while my time shrinks. But... eventually.

**Comment by asymmetric** [comment]
FYI, in Firefox, if I type `/` the "Quick Find" window opens and I can then cycle through results with `Ctrl+G`.

**Comment by lydell** [comment]
(For the record, the VimFx shortcuts where `n`/`N`, not `n`/`p`.)

**Comment by johan-bjareholt** [comment]
Yeah, the only thing i miss now when switching from VimFX is the nN/pP, Ctrl+G is a horrible keybind.
I am constantly creating new windows by mistake when pressing "n" after a search currently lol, it doesn't have to be n/p, but the standard Ctrl+G is horrible so some replacement would be awesome.

**Comment by codehearts** [comment]
I think I'll take a stab at implementing this, I'm sorely missing `n/N` as well coming from VimFx. I suppose `n/N` should be set for the Vimium profile and the other two should have their defaults blank, unless there's an intuitive keybinding that could be used.

**Comment by eejdoowad** [comment]
Definitely take a look at [cVim's](https://github.com/1995eaton/chromium-vim/blob/master/content_scripts/find.js), and [Vimium's](https://github.com/philc/vimium/blob/master/content_scripts/mode_find.coffee) implementation before you start. VimFx was able to use  (now depracated) APIs for Firefox's native search bar.

**Comment by codehearts** [comment]
Mozilla has a [browser.find API](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/API/find) which is only supported by Firefox 57. I may try polyfilling that so a native API may be used in the future.

**Comment by eejdoowad** [comment]
Good plan. Ship a  Firefox implementation using those APIs then later polyfill them on Chrome.

**Comment by codehearts** [comment]
So far I've got a search field popping up in the bottom left when you hit `/`. Search terms use smart case (case-insensitive unless they contain an uppercase), and you can use regex by surrounding the term with `/` (with an optional `i` flag for case-insensitivity). Does that sound ok?

I didn't use `browser.find` in the hopes that doing it manually will support Chrome as well. On the Wikipedia article for Sweden this implementation took ~280ms to find every "i" on the page, as a simple test. That can hopefully be reigned in.

All that's needed is to handle iframes, render matches, stye everything, and implement findNext/findPrev.

**Comment by codehearts** [comment]
I'm not sure if I'll have time to complete this feature anytime soon. These are the issues I'm having:

- I'm not sure how to make the search input open only in the topmost frame
- The search input is affected by styling on webpages
- Each client has an array of `DOMRect`, but I'm unable to send those to the background page. mosi `get` times out, but succeeds if I send a simpler datatype (array length, for example)
- Instead of modifying the underlying webpage, I'm drawing fixed-position rectangles to mimic the unsupported `browser.find` API. This is annoying because
  - These rectangles become out of place if the DOM changes
  - These rectangles appear over elements covering matches on the page (sticky headers)
  - It's hard to make the highlighting look good
    - I tried CSS blend modes to give the appearance of a background color on the text, but applying a blend mode to `guiRoot` affects hints. It also depends on the colors of the underlying webpage
    - I tried darkening the page and using blend modes to "cut out" the matches, this doesn't work with the z-index set on `guiRoot`
    - The `browser.find` API provides text for each rectangle, but it looks off unless you copy the text styling from the underlying page
      - There's no easy way to get that text without the API afaik, because of text wrapping
- DOM ordering is not always the order nodes are displayed on the page, so next/prev match selection is tricky

I originally thought I could port cVim's search implementation (https://github.com/1995eaton/chromium-vim/blob/master/content_scripts/find.js), but they modify the DOM directly and I wanted to bake their regex/case sensitivity options into the search field directly

**Comment by eejdoowad** [comment]
Thanks for the effort @codehearts. This is a preliminary answer I had written before I realized browser.find.find() doesn't support regexes. I'll hopefully get with an update soon.

   I'm not sure if I'll have time to complete this feature anytime soon. These are the issues I'm having:

> - I'm not sure how to make the search input open only in the topmost frame

In the client, send a message to the background component. In the background component, use mosi's `meta` function to query the tab id. Then send a message to the top frame for that tab id.  See [1](https://github.com/lusakasa/saka-key/blob/b87a6b61a2b9393ca95ec525491db5118269ddf2/src/modes/hints/background/index.js#L15) and [2](https://github.com/eejdoowad/mosi#msg---sends-a-message-to-target-nodes-1). When the top frame receives the message, it should create an iframe containing the search input.

> - The search input is affected by styling on webpages

The search input should be inserted into an iframe mounted to the top frame. This is both to isolate it from page styles and to avoid leaking search text/settings to web pages.

> - Each client has an array of `DOMRect`, but I'm unable to send those to the background page. mosi `get` times out, but succeeds if I send a simpler datatype (array length, for example)

Why are DOMRects sent from the client to the background page? Shouldn't it be the opposite? You should rely on Firefox's browser.find.highlightResults APIs to perform highlighting of all matches. You should highlight the active match manually.  Because you only need to message info about a single match, the message payload should be small, and timeouts shouldn't be an issue.

A. Search Term Changes
  1. User changes search input text
  2. New search term is messaged to background page
  3. Background page gets search term from message
  4. Background page calls browser.find.find(searchterm) and caches the result
  5. Background page calls browser.find.highlightResults() and Firefox takes care of highlighting

B. Next or Previous match
  1. User triggers next match
  2. Message is sent to background page
  3. Background page uses result from A.4 to determine next match. It then sends info needed for highlighting to frame containing match. It also sends an "unhighlight" message to frame containing previous match.
  4. Client in frame containing previous match receives message and unhilights old match
  4. Client in frame containing match receives message and highlights active match such that it can be distinguished from other matches (A.5)

C. Search Stops
  1. User closes search input, so a message is sent to the background page
  2. Background page gets this message, then calls browser.find.removeHighlighting()

> - Instead of modifying the underlying webpage, I'm drawing fixed-position rectangles to mimic the unsupported `browser.find` API. This is annoying because
>   - These rectangles become out of place if the DOM changes
>   - These rectangles appear over elements covering matches on the page (sticky headers)
>   - It's hard to make the highlighting look good
>     - I tried CSS blend modes to give the appearance of a background color on the text, but applying a blend mode to `guiRoot` affects hints. It also depends on the colors of the underlying webpage
>     - I tried darkening the page and using blend modes to "cut out" the matches, this doesn't work with the z-index set on `guiRoot`
>     - The `browser.find` API provides text for each rectangle, but it looks off unless you copy the text styling from the underlying page
      - There's no easy way to get that text without the API afaik, because of text wrapping

Note: this only applies to Chrome, which lacks browser.find.*. Stick to modifying the DOM directly like cVim. It's a much, much, much easier problem. It's extremely hard to keep highlighting in sync with text if the highlighting rects are mounted to an unrelated node. Corner cases abound.

> - DOM ordering is not always the order nodes are displayed on the page, so next/prev match selection is tricky

Use the ordering from A.4

> I originally thought I could port [cVim's search implementation](https://github.com/1995eaton/chromium-vim/blob/master/content_scripts/find.js), but they modify the DOM directly and I wanted to bake their regex/case sensitivity options into the search field directly. I have a [Find mode branch](https://github.com/codehearts/saka-key/tree/find-mode) with all of my work so far

Should have read this earlier...
